File: src/01-easy/moves-in-maze/solution.rb
STDOUT.sync = true
C = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
w, h = gets.split.map(&:to_i)
g = h.times.map { gets.chomp }
d = Array.new(h) { Array.new(w, -1) }
sr = sc = 0
h.times { |r| w.times { |c| (sr, sc = r, c) if g[r][c] == 'S' } }
d[sr][sc] = 0
q = [[sr, sc]]
while q.any?
    r, c = q.shift
    [[0,1],[0,-1],[1,0],[-1,0]].each do |dr, dc|
        nr, nc = (r+dr) % h, (c+dc) % w
        if g[nr][nc] != '#' && d[nr][nc] == -1
            d[nr][nc] = d[r][c] + 1; q << [nr, nc]
        end
    end
end
h.times { |r| puts (0...w).map { |c| g[r][c] == '#' ? '#' : d[r][c] == -1 ? '.' : C[d[r][c]] }.join }

File: src/01-easy/moves-in-maze/solution.pl
$| = 1;
my $C = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
chomp(my $line = <STDIN>); my ($w, $h) = split / /, $line;
my @g; for (1..$h) { chomp(my $l = <STDIN>); push @g, $l; }
my @d; for my $r (0..$h-1) { for my $c (0..$w-1) { $d[$r][$c] = -1; } }
my ($sr, $sc) = (0, 0);
for my $r (0..$h-1) { for my $c (0..$w-1) { ($sr,$sc)=($r,$c) if substr($g[$r],$c,1) eq 'S'; } }
$d[$sr][$sc] = 0;
my @q = ([$sr, $sc]);
my @dr = (0,0,1,-1); my @dc = (1,-1,0,0);
while (@q) {
    my ($r, $c) = @{shift @q};
    for my $i (0..3) {
        my $nr = ($r+$dr[$i]+$h)%$h; my $nc = ($c+$dc[$i]+$w)%$w;
        if (substr($g[$nr],$nc,1) ne '#' && $d[$nr][$nc] == -1) { $d[$nr][$nc] = $d[$r][$c]+1; push @q, [$nr,$nc]; }
    }
}
for my $r (0..$h-1) {
    print join('', map { substr($g[$r],$_,1) eq '#' ? '#' : $d[$r][$_]==-1 ? '.' : substr($C,$d[$r][$_],1) } 0..$w-1), "\n";
}
